src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java

```diff
@@ -1004,10 +1018,20 @@ public Label getFdoPrefetchHintsLabel() {
   /** See {@link #targetLibcTopLabel} documentation. * */
   @Override
   public FragmentOptions getNormalized() {
+    CppOptions newOptions = (CppOptions) this.clone();
+    boolean changed = false;
     if (targetLibcTopLabel != null
         && targetLibcTopLabel.getName().equals(TARGET_LIBC_TOP_NOT_YET_SET)) {
-      CppOptions newOptions = (CppOptions) this.clone();
       newOptions.targetLibcTopLabel = libcTopLabel;
+      changed = true;
+    }
+    if (hostCrosstoolTop != null
+        && hostCrosstoolTop.getName().equals(HOST_CROSSTOOL_TOP_NOT_YET_SET)) {
```

Comment:
Do we actually need a special marker value or does null still work here? As the code stands now, what case does the `!= null` check catch?

It took me a lot of time to follow the logic from https://github.com/bazelbuild/bazel/commit/6720ff31ba45b6225993cf7a0ccf50ad2001f0e9, which set this basic pattern. Even with the crucial comments at https://github.com/bazelbuild/bazel/blob/6720ff31ba45b6225993cf7a0ccf50ad2001f0e9/src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java#L653-L656.

My recollection now is that null default doesn't work for `targetLibcTopLabel` because the similarly named `libcTopLabel` does default to null. So if you build in a default target configuration but with [`--android_grte_top`](https://github.com/bazelbuild/bazel/blob/18ee6bacdb079432d71bbcbefeb5da8586dbd05f/src/main/java/com/google/devtools/build/lib/rules/android/AndroidConfiguration.java#L235)=`<something non-null>`, the top-level version of `getNormalized` would set `targetLibcTopLabel = libcTopLabel = null`, and then the Android version would re-trigger because it thought `null` meant "never set".

`host_crosstool_top` isn't quite the same, is it? Because the equivalent to `libcTopLabel` is `crosstoolTop`, which itself doesn't default to null? 

Whatever logic we want to use, can we document these obscure subtleties super-precisely?

Reply:
You're right, using "null" is fine for host_crosstool_top. Let me know if the current documentation is insufficient.